README.md
# mic1-asm README

Language extension for the MIC1 assembly language

## Features

Describe specific features of your extension including screenshots of your extension in action. Image paths are relative to this README file.

For example if there is an image subfolder under your extension project workspace:

\!\[feature X\]\(images/feature-x.png\)

> Tip: Many popular extensions utilize animations. This is an excellent way to show off your extension! We recommend short, focused animations that are easy to follow.

## Requirements

If you have any requirements or dependencies, add a section describing those and how to install and configure them.

## Extension Settings

Include if your extension adds any VS Code settings through the `contributes.configuration` extension point.

For example:

This extension contributes the following settings:

* `myExtension.enable`: Enable/disable this extension.
* `myExtension.thing`: Set to `blah` to do something.

## Known Issues

Calling out known issues can help limit users opening duplicate issues against your extension.

## Release Notes

Users appreciate release notes as you update your extension.

### 1.0.0

Initial release of ...

### 1.0.1

Fixed issue #.

### 1.1.0

Added features X, Y, and Z.

---

## Working with Markdown

You can author your README using Visual Studio Code. Here are some useful editor keyboard shortcuts:

* Split the editor (`Cmd+\` on macOS or `Ctrl+\` on Windows and Linux).
* Toggle preview (`Shift+Cmd+V` on macOS or `Shift+Ctrl+V` on Windows and Linux).
* Press `Ctrl+Space` (Windows, Linux, macOS) to see a list of Markdown snippets.

## For more information

* [Visual Studio Code's Markdown Support](http://code.visualstudio.com/docs/languages/markdown)
* [Markdown Syntax Reference](https://help.github.com/articles/markdown-basics/)

**Enjoy!**


CHANGELOG.md
# Change Log

All notable changes to the "mic1-asm" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

out/main.js
"use strict";var d=Object.create;var c=Object.defineProperty;var g=Object.getOwnPropertyDescriptor;var l=Object.getOwnPropertyNames;var w=Object.getPrototypeOf,m=Object.prototype.hasOwnProperty;var u=(x,s)=>{for(var a in s)c(x,a,{get:s[a],enumerable:!0})},i=(x,s,a,o)=>{if(s&&typeof s=="object"||typeof s=="function")for(let t of l(s))!m.call(x,t)&&t!==a&&c(x,t,{get:()=>s[t],enumerable:!(o=g(s,t))||o.enumerable});return x};var h=(x,s,a)=>(a=x!=null?d(w(x)):{},i(s||!x||!x.__esModule?c(a,"default",{value:x,enumerable:!0}):a,x)),f=x=>i(c({},"__esModule",{value:!0}),x);var b={};u(b,{activate:()=>y,deactivate:()=>S});module.exports=f(b);var r=h(require("vscode"));function y(x){console.log("activated"),x.subscriptions.push(v)}var v=r.languages.registerHoverProvider({scheme:"file",language:"asm"},{provideHover(x,s,a){let o=x.getWordRangeAtPosition(s),t=x.getText(o).toLowerCase();if(N(t)&&p.has(t))return new r.Hover(p.get(t).markdownDesc)}}),j=["lodd","stod","addd","subd","jpos","jzer","jump","loco","lodl","stol","addl","subl","jneg","jnze","call","pshi","popi","push","pop","retn","swap","insp","desp","halt"];function N(x){return j.indexOf(x)>=0}var e=class{constructor(s,a){this.argName=s,this.xMode=a,this.desc=`<${this.argName}: ${this.xMode}>`}},n=class{constructor(s,a,o){this.name=s,this.signature=o,this.description=a,this.signature?.arg1&&(this.description=this.description.replace("$1",String.raw`\<${this.signature.arg1.argName}\>`)),this.signature?.arg2&&(this.description=this.description.replace("$2",String.raw`\<${this.signature.arg2.argName}\>`));let t=this.name;this.signature?.arg1&&(t+=` ${this.signature?.arg1.desc}`),this.signature?.arg2&&(t+=` ${this.signature?.arg2.desc}`),this.markdownDesc=new r.MarkdownString().appendCodeblock(t,"asm").appendMarkdown(`
 --- 
`).appendMarkdown(this.description)}},p=new Map([["lodd",new n("lodd","Load direct\n\n`ac := m[x]`\n\n`0000xxxxxxxxxxxx`",{arg1:new e("x","address")})],["stod",new n("stod","Store direct\n\n`m[x] := ac`\n\n`0001xxxxxxxxxxxx`",{arg1:new e("x","address")})],["addd",new n("addd","Add direct\n\n`ac := ac + m[x]`\n\n`0010xxxxxxxxxxxx`",{arg1:new e("x","address")})],["subd",new n("subd","Subtract direct\n\n`ac := ac - m[x]`\n\n`0011xxxxxxxxxxxx`",{arg1:new e("x","address")})],["jpos",new n("jpos","Jump positive\n\n`if ac >= 0 then pc := x`\n\n`0100xxxxxxxxxxxx`",{arg1:new e("x","location")})],["jzer",new n("jzer","Jump zero\n\n`if ac = 0 then pc := x`\n\n`0101xxxxxxxxxxxx`",{arg1:new e("x","location")})],["jump",new n("jump","Jump\n\n`pc := x`\n\n`0110xxxxxxxxxxxx`",{arg1:new e("x","location")})],["loco",new n("loco","Load constant\n\n`ac := x (0 <= x <= 4095)`\n\n`0111xxxxxxxxxxxx`",{arg1:new e("x","constant")})],["lodl",new n("lodl","Load local\n\n`ac := m[sp + x]`\n\n`1000xxxxxxxxxxxx`",{arg1:new e("x","offset")})],["stol",new n("stol","Store local\n\n`m[x + sp] := ac`\n\n`1001xxxxxxxxxxxx`",{arg1:new e("x","offset")})],["addl",new n("addl","Add local\n\n`ac := ac + m[sp + x]`\n\n`1010xxxxxxxxxxxx`",{arg1:new e("x","offset")})],["subl",new n("subl","Subtract local\n\n`ac := ac - m[sp + x]`\n\n`1011xxxxxxxxxxxx`",{arg1:new e("x","offset")})],["jneg",new n("jneg","Jump negative\n\n`if ac < 0 then pc := x`\n\n`1100xxxxxxxxxxxx`",{arg1:new e("x","location")})],["jnze",new n("jnze","Jump nonzero\n\n`if ac != 0 then pc := x`\n\n`1101xxxxxxxxxxxx`",{arg1:new e("x","location")})],["call",new n("call","Call procedure\n\n`sp := sp - 1;`\n\n`m[sp] := pc;`\n\n`pc := x`\n\n`1110xxxxxxxxxxxx`",{arg1:new e("x","location")})],["pshi",new n("pshi","Push indirect\n\n`sp := sp - 1;`\n\n`m[sp] := m[ac]`\n\n`1111000000000000`")],["popi",new n("popi","Pop indirect\n\n`m[ac] := m[sp];`\n\n`sp := sp + 1`\n\n`1111001000000000`")],["push",new n("push","Push onto stack\n\n`sp := sp - 1;`\n\n`m[sp] := ac`\n\n`1111010000000000`")],["pop",new n("pop","Pop from stack\n\n`ac := m[sp];`\n\n`sp := sp + 1`\n\n`1111011000000000`")],["retn",new n("retn","Return\n\n`pc := m[sp];`\n\n`sp := sp + 1`\n\n`1111100000000000`")],["swap",new n("swap","Swap ac, sp\n\n`tmp := ac;`\n\n`ac := sp;`\n\n`sp := tmp`\n\n`1111101000000000`")],["insp",new n("insp","Increment sp\n\n`sp := sp + y (0 <= y <= 255)`\n\n`11111100yyyyyyyy`",{arg1:new e("y","constant")})],["desp",new n("desp","Decrement sp\n\n`sp := sp - y (0 <= y <= 255)`\n\n`11111110yyyyyyyy`")],["halt",new n("halt","Go to debugger interface\n\n`1111111100000000`")]]);function S(){}0&&(module.exports={activate,deactivate});


src/extension.ts
import * as vscode from "vscode";

export function activate(context: vscode.ExtensionContext) {
	console.log('activated');
	
	context.subscriptions.push(HOVER_DEFN);
}

const HOVER_DEFN = vscode.languages.registerHoverProvider(
	{ scheme: "file", language: "asm" },
	{
		provideHover(document, position, token) {
			const range = document.getWordRangeAtPosition(position);
			const word = document.getText(range).toLowerCase();
			if(isOpcodeName(word) && OPCODES.has(word)) {
				return new vscode.Hover(OPCODES.get(word)!.markdownDesc);
			}
		},
	},
);

// address - m[x]
// location - pc := x
// constant - ac := x
// offset - m[sp + x]
type XMode = "address" | "location" | "constant" | "offset"

const opcodeNames = [
	"lodd",
	"stod",
	"addd",
	"subd",
	"jpos",
	"jzer",
	"jump",
	"loco",
	"lodl",
	"stol",
	"addl",
	"subl",
	"jneg",
	"jnze",
	"call",
	"pshi",
	"popi",
	"push",
	"pop",
	"retn",
	"swap",
	"insp",
	"desp",
	"halt",
] as const;

type OpcodeName = typeof opcodeNames[number];
function isOpcodeName(x: string): x is OpcodeName {
	return (opcodeNames as readonly string[]).indexOf(x) >= 0;
}

type Signature = {
	arg1?: Argument;
	arg2?: Argument;
};

class Argument {
	argName: string;
	xMode: XMode;
	desc: string;
	constructor(argName: string, mode: XMode) {
		this.argName = argName;
		this.xMode = mode;
		this.desc = `<${this.argName}: ${this.xMode}>`;
	}
};

class Opcode {
	name: OpcodeName;
	signature?: Signature;
	description: string;
	markdownDesc: vscode.MarkdownString;
	
	constructor(name: OpcodeName, description: string, signature?: Signature) {
		this.name = name;
		this.signature = signature;
		this.description = description;
		// Format description from signature
		if (this.signature?.arg1) {
			this.description = this.description.replace("$1", String.raw`\<${this.signature.arg1.argName}\>`);
		}
		if (this.signature?.arg2) {
			this.description = this.description.replace("$2", String.raw`\<${this.signature.arg2.argName}\>`);
		}
		
		// Generate markdownDesc
		let heading = this.name;
		if (this.signature?.arg1) {
			heading += ` ${this.signature?.arg1.desc}`;
		}
		if (this.signature?.arg2) {
			heading += ` ${this.signature?.arg2.desc}`;
		}
		this.markdownDesc = new vscode.MarkdownString()
			.appendCodeblock(heading, "asm")
			.appendMarkdown("\n --- \n") // insert horizontal rule
			.appendMarkdown(this.description);
	}
}

const OPCODES = new Map<OpcodeName, Opcode>([
	[
		"lodd",
		new Opcode("lodd", "Load direct\n\n`ac := m[x]`\n\n`0000xxxxxxxxxxxx`", {
			arg1: new Argument("x", "address")
		})
	],
	[
		"stod",
		new Opcode("stod", "Store direct\n\n`m[x] := ac`\n\n`0001xxxxxxxxxxxx`", {
			arg1: new Argument("x", "address")
		})
	],
	[
		"addd",
		new Opcode("addd", "Add direct\n\n`ac := ac + m[x]`\n\n`0010xxxxxxxxxxxx`", {
			arg1: new Argument("x", "address")
		})
	],
	[
		"subd",
		new Opcode("subd", "Subtract direct\n\n`ac := ac - m[x]`\n\n`0011xxxxxxxxxxxx`", {
			arg1: new Argument("x", "address")
		})
	],
	[
		"jpos",
		new Opcode("jpos", "Jump positive\n\n`if ac >= 0 then pc := x`\n\n`0100xxxxxxxxxxxx`", {
			arg1: new Argument("x", "location")
		})
	],
	[
		"jzer",
		new Opcode("jzer", "Jump zero\n\n`if ac = 0 then pc := x`\n\n`0101xxxxxxxxxxxx`", {
			arg1: new Argument("x", "location")
		})
	],
	[
		"jump",
		new Opcode("jump", "Jump\n\n`pc := x`\n\n`0110xxxxxxxxxxxx`", {
			arg1: new Argument("x", "location")
		})
	],
	[
		"loco",
		new Opcode("loco", "Load constant\n\n`ac := x (0 <= x <= 4095)`\n\n`0111xxxxxxxxxxxx`", {
			arg1: new Argument("x", "constant")
		})
	],
	[
		"lodl",
		new Opcode("lodl", "Load local\n\n`ac := m[sp + x]`\n\n`1000xxxxxxxxxxxx`", {
			arg1: new Argument("x", "offset")
		})
	],
	[
		"stol",
		new Opcode("stol", "Store local\n\n`m[x + sp] := ac`\n\n`1001xxxxxxxxxxxx`", {
			arg1: new Argument("x", "offset")
		})
	],
	[
		"addl",
		new Opcode("addl", "Add local\n\n`ac := ac + m[sp + x]`\n\n`1010xxxxxxxxxxxx`", {
			arg1: new Argument("x", "offset")
		})
	],
	[
		"subl",
		new Opcode("subl", "Subtract local\n\n`ac := ac - m[sp + x]`\n\n`1011xxxxxxxxxxxx`", {
			arg1: new Argument("x", "offset")
		})
	],
	[
		"jneg",
		new Opcode("jneg", "Jump negative\n\n`if ac < 0 then pc := x`\n\n`1100xxxxxxxxxxxx`", {
			arg1: new Argument("x", "location")
		})
	],
	[
		"jnze",
		new Opcode("jnze", "Jump nonzero\n\n`if ac != 0 then pc := x`\n\n`1101xxxxxxxxxxxx`", {
			arg1: new Argument("x", "location")
		})
	],
	[
		"call",
		new Opcode("call", "Call procedure\n\n`sp := sp - 1;`\n\n`m[sp] := pc;`\n\n`pc := x`\n\n`1110xxxxxxxxxxxx`", {
			arg1: new Argument("x", "location")
		})
	],
	[
		"pshi",
		new Opcode("pshi", "Push indirect\n\n`sp := sp - 1;`\n\n`m[sp] := m[ac]`\n\n`1111000000000000`")
	],
	[
		"popi",
		new Opcode("popi", "Pop indirect\n\n`m[ac] := m[sp];`\n\n`sp := sp + 1`\n\n`1111001000000000`")
	],
	[
		"push",
		new Opcode("push", "Push onto stack\n\n`sp := sp - 1;`\n\n`m[sp] := ac`\n\n`1111010000000000`")
	],
	[
		"pop",
		new Opcode("pop", "Pop from stack\n\n`ac := m[sp];`\n\n`sp := sp + 1`\n\n`1111011000000000`")
	],
	[
		"retn",
		new Opcode("retn", "Return\n\n`pc := m[sp];`\n\n`sp := sp + 1`\n\n`1111100000000000`")
	],
	[
		"swap",
		new Opcode("swap", "Swap ac, sp\n\n`tmp := ac;`\n\n`ac := sp;`\n\n`sp := tmp`\n\n`1111101000000000`")
	],
	[
		"insp",
		new Opcode("insp", "Increment sp\n\n`sp := sp + y (0 <= y <= 255)`\n\n`11111100yyyyyyyy`", {
			arg1: new Argument("y", "constant")
		})
	],
	[
		"desp",
		new Opcode("desp", "Decrement sp\n\n`sp := sp - y (0 <= y <= 255)`\n\n`11111110yyyyyyyy`")
	],
	[
		"halt",
		new Opcode("halt", "Go to debugger interface\n\n`1111111100000000`")
	],
]);

export function deactivate() {}


out/extension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deactivate = exports.activate = void 0;
const vscode = require("vscode");
function activate(context) {
    console.log('activated');
    context.subscriptions.push(HOVER_DEFN);
}
exports.activate = activate;
const HOVER_DEFN = vscode.languages.registerHoverProvider({ scheme: "file", language: "asm" }, {
    provideHover(document, position, token) {
        const range = document.getWordRangeAtPosition(position);
        const word = document.getText(range).toLowerCase();
        if (isOpcodeName(word) && OPCODES.has(word)) {
            return new vscode.Hover(OPCODES.get(word).markdownDesc);
        }
    },
});
const opcodeNames = [
    "lodd",
    "stod",
    "addd",
    "subd",
    "jpos",
    "jzer",
    "jump",
    "loco",
    "lodl",
    "stol",
    "addl",
    "subl",
    "jneg",
    "jnze",
    "call",
    "pshi",
    "popi",
    "push",
    "pop",
    "retn",
    "swap",
    "insp",
    "desp",
    "halt",
];
function isOpcodeName(x) {
    return opcodeNames.indexOf(x) >= 0;
}
class Argument {
    constructor(argName, mode) {
        this.argName = argName;
        this.xMode = mode;
        this.desc = `<${this.argName}: ${this.xMode}>`;
    }
}
;
class Opcode {
    constructor(name, description, signature) {
        this.name = name;
        this.signature = signature;
        this.description = description;
        // Format description from signature
        if (this.signature?.arg1) {
            this.description = this.description.replace("$1", String.raw `\<${this.signature.arg1.argName}\>`);
        }
        if (this.signature?.arg2) {
            this.description = this.description.replace("$2", String.raw `\<${this.signature.arg2.argName}\>`);
        }
        // Generate markdownDesc
        let heading = this.name;
        if (this.signature?.arg1) {
            heading += ` ${this.signature?.arg1.desc}`;
        }
        if (this.signature?.arg2) {
            heading += ` ${this.signature?.arg2.desc}`;
        }
        this.markdownDesc = new vscode.MarkdownString()
            .appendCodeblock(heading, "asm")
            .appendMarkdown("\n --- \n") // insert horizontal rule
            .appendMarkdown(this.description);
    }
}
const OPCODES = new Map([
    [
        "lodd",
        new Opcode("lodd", "Load direct\n\n`ac := m[x]`\n\n`0000xxxxxxxxxxxx`", {
            arg1: new Argument("x", "address")
        })
    ],
    [
        "stod",
        new Opcode("stod", "Store direct\n\n`m[x] := ac`\n\n`0001xxxxxxxxxxxx`", {
            arg1: new Argument("x", "address")
        })
    ],
    [
        "addd",
        new Opcode("addd", "Add direct\n\n`ac := ac + m[x]`\n\n`0010xxxxxxxxxxxx`", {
            arg1: new Argument("x", "address")
        })
    ],
    [
        "subd",
        new Opcode("subd", "Subtract direct\n\n`ac := ac - m[x]`\n\n`0011xxxxxxxxxxxx`", {
            arg1: new Argument("x", "address")
        })
    ],
    [
        "jpos",
        new Opcode("jpos", "Jump positive\n\n`if ac >= 0 then pc := x`\n\n`0100xxxxxxxxxxxx`", {
            arg1: new Argument("x", "location")
        })
    ],
    [
        "jzer",
        new Opcode("jzer", "Jump zero\n\n`if ac = 0 then pc := x`\n\n`0101xxxxxxxxxxxx`", {
            arg1: new Argument("x", "location")
        })
    ],
    [
        "jump",
        new Opcode("jump", "Jump\n\n`pc := x`\n\n`0110xxxxxxxxxxxx`", {
            arg1: new Argument("x", "location")
        })
    ],
    [
        "loco",
        new Opcode("loco", "Load constant\n\n`ac := x (0 <= x <= 4095)`\n\n`0111xxxxxxxxxxxx`", {
            arg1: new Argument("x", "constant")
        })
    ],
    [
        "lodl",
        new Opcode("lodl", "Load local\n\n`ac := m[sp + x]`\n\n`1000xxxxxxxxxxxx`", {
            arg1: new Argument("x", "offset")
        })
    ],
    [
        "stol",
        new Opcode("stol", "Store local\n\n`m[x + sp] := ac`\n\n`1001xxxxxxxxxxxx`", {
            arg1: new Argument("x", "offset")
        })
    ],
    [
        "addl",
        new Opcode("addl", "Add local\n\n`ac := ac + m[sp + x]`\n\n`1010xxxxxxxxxxxx`", {
            arg1: new Argument("x", "offset")
        })
    ],
    [
        "subl",
        new Opcode("subl", "Subtract local\n\n`ac := ac - m[sp + x]`\n\n`1011xxxxxxxxxxxx`", {
            arg1: new Argument("x", "offset")
        })
    ],
    [
        "jneg",
        new Opcode("jneg", "Jump negative\n\n`if ac < 0 then pc := x`\n\n`1100xxxxxxxxxxxx`", {
            arg1: new Argument("x", "location")
        })
    ],
    [
        "jnze",
        new Opcode("jnze", "Jump nonzero\n\n`if ac != 0 then pc := x`\n\n`1101xxxxxxxxxxxx`", {
            arg1: new Argument("x", "location")
        })
    ],
    [
        "call",
        new Opcode("call", "Call procedure\n\n`sp := sp - 1;`\n\n`m[sp] := pc;`\n\n`pc := x`\n\n`1110xxxxxxxxxxxx`", {
            arg1: new Argument("x", "location")
        })
    ],
    [
        "pshi",
        new Opcode("pshi", "Push indirect\n\n`sp := sp - 1;`\n\n`m[sp] := m[ac]`\n\n`1111000000000000`")
    ],
    [
        "popi",
        new Opcode("popi", "Pop indirect\n\n`m[ac] := m[sp];`\n\n`sp := sp + 1`\n\n`1111001000000000`")
    ],
    [
        "push",
        new Opcode("push", "Push onto stack\n\n`sp := sp - 1;`\n\n`m[sp] := ac`\n\n`1111010000000000`")
    ],
    [
        "pop",
        new Opcode("pop", "Pop from stack\n\n`ac := m[sp];`\n\n`sp := sp + 1`\n\n`1111011000000000`")
    ],
    [
        "retn",
        new Opcode("retn", "Return\n\n`pc := m[sp];`\n\n`sp := sp + 1`\n\n`1111100000000000`")
    ],
    [
        "swap",
        new Opcode("swap", "Swap ac, sp\n\n`tmp := ac;`\n\n`ac := sp;`\n\n`sp := tmp`\n\n`1111101000000000`")
    ],
    [
        "insp",
        new Opcode("insp", "Increment sp\n\n`sp := sp + y (0 <= y <= 255)`\n\n`11111100yyyyyyyy`", {
            arg1: new Argument("y", "constant")
        })
    ],
    [
        "desp",
        new Opcode("desp", "Decrement sp\n\n`sp := sp - y (0 <= y <= 255)`\n\n`11111110yyyyyyyy`")
    ],
    [
        "halt",
        new Opcode("halt", "Go to debugger interface\n\n`1111111100000000`")
    ],
]);
function deactivate() { }
exports.deactivate = deactivate;
//# sourceMappingURL=extension.js.map